pkg/controller/controller_ref_manager.go

```diff
@@ -0,0 +1,168 @@
+	}
+
+	return m.ClaimObject(rc, match, adopt, release)
+}
+
+// ClaimReplicationControllers tries to take ownership of a list of ReplicationControllers.
+//
+// It will reconcile the following:
+//   * Adopt orphans if the selector matches.
+//   * Release owned objects if the selector no longer matches.
+//
+// A non-nil error is returned if some form of reconciliation was attempted and
+// failed. Usually, controllers should try again later in case reconciliation
+// is still needed.
+//
+// If the error is nil, either the reconciliation succeeded, or no
+// reconciliation was necessary. The list of ReplicationControllers that you now own is
+// returned.
+func (m *RCControllerRefManager) ClaimReplicationControllers(rcs []*kapi.ReplicationController) ([]*kapi.ReplicationController, error) {
+	var claimed []*kapi.ReplicationController
+	var errlist []error
+
+	for _, rc := range rcs {
+		ok, err := m.ClaimReplicationController(rc)
+		if err != nil {
+			errlist = append(errlist, err)
+			continue
+		}
+		if ok {
+			claimed = append(claimed, rc)
+		}
+	}
+	return claimed, kutilerrors.NewAggregate(errlist)
+}
+
+// AdoptReplicationController sends a patch to take control of the ReplicationController. It returns the error if
+// the patching fails.
+func (m *RCControllerRefManager) AdoptReplicationController(rs *kapi.ReplicationController) error {
+	if err := m.CanAdopt(); err != nil {
+		return fmt.Errorf("can't adopt ReplicationController %v/%v (%v): %v", rs.Namespace, rs.Name, rs.UID, err)
+	}
+	// Note that ValidateOwnerReferences() will reject this patch if another
+	// OwnerReference exists with controller=true.
+	addControllerPatch := fmt.Sprintf(
+		`{"metadata":{"ownerReferences":[{"apiVersion":"%s","kind":"%s","name":"%s","uid":"%s","controller":true,"blockOwnerDeletion":true}],"uid":"%s"}}`,
+		m.controllerKind.GroupVersion(), m.controllerKind.Kind,
+		m.Controller.GetName(), m.Controller.GetUID(), rs.UID)
+	return m.rcControl.PatchReplicationController(rs.Namespace, rs.Name, []byte(addControllerPatch))
+}
+
+// ReleaseReplicationController sends a patch to free the ReplicationController from the control of the Deployment controller.
+// It returns the error if the patching fails. 404 and 422 errors are ignored.
+func (m *RCControllerRefManager) ReleaseReplicationController(rc *kapi.ReplicationController) error {
+	glog.V(2).Infof("patching ReplicationController %s_%s to remove its controllerRef to %s/%s:%s",
+		rc.Namespace, rc.Name, m.controllerKind.GroupVersion(), m.controllerKind.Kind, m.Controller.GetName())
+	deleteOwnerRefPatch := fmt.Sprintf(`{"metadata":{"ownerReferences":[{"$patch":"delete","uid":"%s"}],"uid":"%s"}}`, m.Controller.GetUID(), rc.UID)
+	err := m.rcControl.PatchReplicationController(rc.Namespace, rc.Name, []byte(deleteOwnerRefPatch))
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			// If the ReplicationController no longer exists, ignore it.
+			return nil
+		}
+		if kerrors.IsInvalid(err) {
```

Comment:
@mfojtik essentially this whole file is from upstream with a bit of `sed s/RS/RC/g`. Do we want to deviate here w.r.t. future diffs?